### PR TITLE
Terraform updated for DHCP based NSX-T segment

### PR DESCRIPTION
**Pull request by Pushpalatha555**
**validated the Terraform changes for both DHCP/static**

### CODE REVIEW

File: examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    id = "shared"
+  }
+  resource_permissions {
+    all = true
+  }
+  dhcp_network{
+    dhcp_enabled = true
+    dhcp_config {
+      dhcp_type = "dhcpLocal"
+      dhcp_server = "/infra/dhcp-server-configs/89e6b52f-2489-41b5-96f2-3393eaaad4b9"
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf:24]: can you please add datasource here  for dhcp_server ?

File: examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  scope_id     = data.hpegl_vmaas_transport_zone.tf_zone.provider_id
+  cidr         = "193.2.3.1/20" 
+  primary_dns = "8.8.8.8"
+  scan_network = false
+  active       = true
+  allow_static_override = true
+  appliance_url_proxy_bypass = true
+  group {
+    id = "shared"
+  }
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf:16]: group is not right, please put back in the old way as group_id

File: examples/resources/hpegl_vmaas_network/nsx_t_static_network.tf [hunk 3]
```diff
@@ -1,20 +1,25 @@
...
-  cidr         = "168.72.10.1/18"
+  cidr         = "193.2.3.1/20"
   gateway      = "168.72.10.1"
   primary_dns  = "8.8.8.8"
   scan_network = false
   pool_id = 7
   active       = true
-  config {
-    connected_gateway = data.hpegl_vmaas_router.tf_router.provider_id
-  }
```
Comment (manjunath-batakurki) [inline examples/resources/hpegl_vmaas_network/nsx_t_static_network.tf:16]: We need to check with Reuben, if we need to maintain compatibility here ? these may break existing automation, consumers if they are using it.

File: internal/cmp/resource_network.go [hunk 4]
```diff
@@ -138,12 +142,24 @@ func (r *resNetwork) Delete(ctx context.Context, d *utils.Data, meta interface{}
...
 
-func alignNetworkReq(request *models.CreateNetwork) {
-	request.Site.ID = request.GroupID
-	if request.NetworkDomainID != 0 {
-		request.NetworkDomain = &models.IDModel{ID: request.NetworkDomainID}
+func (r *resNetwork) networkRequest(createReq *models.CreateNetwork) error {
+	if createReq.TfDhcpNetwork != nil {
+		createReq.DhcpServer = createReq.TfDhcpNetwork.DhcpServer
+		if createReq.TfDhcpNetwork.Config != nil {
+			createReq.Config.ConnectedGateway = createReq.TfDhcpNetwork.Config.ConnectedGateway
```
Comment (manjunath-batakurki) [inline internal/cmp/resource_network.go:149]: please check if each of these fields (by checking UI if these values are shown righly) are set while you are testing it

File: internal/cmp/resource_network.go [hunk 5]
```diff
@@ -138,12 +142,24 @@ func (r *resNetwork) Delete(ctx context.Context, d *utils.Data, meta interface{}
 	return err
 }
 
-func alignNetworkReq(request *models.CreateNetwork) {
-	request.Site.ID = request.GroupID
-	if request.NetworkDomainID != 0 {
-		request.NetworkDomain = &models.IDModel{ID: request.NetworkDomainID}
```
Comment (manjunath-batakurki) [inline internal/cmp/resource_network.go:144]: why did you remove "if request.NetworkDomainID != 0 {
		request.NetworkDomain = &models.IDModel{ID: request.NetworkDomainID}"

File: internal/cmp/resource_network.go [hunk 6]
```diff
@@ -138,12 +142,24 @@ func (r *resNetwork) Delete(ctx context.Context, d *utils.Data, meta interface{}
...
+			createReq.Config.VlanIDs = createReq.TfDhcpNetwork.Config.VlanIDs
+			createReq.Config.SubnetDhcpLeaseTime = createReq.TfDhcpNetwork.Config.SubnetDhcpLeaseTime
+			createReq.Config.SubnetDhcpServerAddress = createReq.TfDhcpNetwork.Config.SubnetDhcpServerAddress
+			createReq.Config.SubnetIPManagementType = createReq.TfDhcpNetwork.Config.SubnetIPManagementType
+			createReq.Config.SubnetIPServerID = createReq.TfDhcpNetwork.Config.SubnetIPServerID
+		}
 	}
-	if request.ProxyID != 0 {
-		request.NetworkProxy = &models.IDModel{ID: request.ProxyID}
+	if createReq.TfStaticNetwork != nil {
```
Comment (manjunath-batakurki) [inline internal/cmp/resource_network.go:157]: why did you remove if request.ProxyID != 0 {
		request.NetworkProxy = &models.IDModel{ID: request.ProxyID}

File: internal/resources/data_source_dhcp_server.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+				Description: "id can be obtained by using DHCP Server datasource/resource.",
+			},
+			"provider_id": {
+				Type:        schema.TypeString,
+				Computed:    true,
+				Description: "ProviderId of the DHCP Server. Use the provider_id  while creating DHCP NSX-T Segment Network",
+			},
+		},
+		ReadContext: DHCPServerReadContext,
+		Description: `The ` + DSDhcpServer + ` data source can be used to discover the ID of a hpegl vmaas dhcpserver.
```
Comment (manjunath-batakurki) [inline internal/resources/data_source_dhcp_server.go:34]: hpegl vmaas dhcpserver. Please DHCP (**All Caps**) in description fields

File: internal/resources/diffValidation/resource_nsxt_network.go [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+			}
+		}
+	}
+	if len((staticTypes).([]interface{})) != 0 {
+		for _, networks := range staticTypes.([]interface{}) {
+			if len(networks.([]interface{})) == 0 {
+				return fmt.Errorf("please provide " + staticConfig + " for the Configuration")
+			}
+		}
+	}
```
Comment (manjunath-batakurki) [inline internal/resources/diffValidation/resource_nsxt_network.go:60]: i see line 47-53 and 53-60 are same, not sure why these should like this . Can u have just one function to do 47-53 for both static and dhcp ?

File: internal/resources/resource_networks.go [hunk 9]
```diff
@@ -30,16 +32,6 @@ func Network() *schema.Resource {
 				Optional:    true,
 				Description: "Display name of the NSX-T network.",
 			},
-			"group_id": {
```
Comment (manjunath-batakurki) [inline internal/resources/resource_networks.go:33]: please dont change group_id

File: internal/resources/resource_networks.go [hunk 10]
```diff
@@ -50,21 +42,6 @@ func Network() *schema.Resource {
...
-			"internal_id": {
-				Type:        schema.TypeString,
-				Computed:    true,
-				Description: "Internal ID of the network",
-			},
-			"unique_id": {
-				Type:        schema.TypeString,
-				Computed:    true,
-				Description: "Unique ID of the network",
-			},
```
Comment (manjunath-batakurki) [inline internal/resources/resource_networks.go:67]: why are u removing these ? are these specific to static network only ?

File: internal/resources/resource_networks.go [hunk 11]
```diff
@@ -138,23 +109,39 @@ func Network() *schema.Resource {
 				Optional:    true,
 				Description: "If set to true, network will allow static override",
 			},
-			"config": {
+			"scope_id": {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: "Transport Zone ID. Use " + DSTransportZone + " Data source's `provider_id` here.",
+			},
+			"group": {
```
Comment (manjunath-batakurki) [inline internal/resources/resource_networks.go:117]: please use group id

File: internal/resources/resource_networks.go [hunk 12]
```diff
@@ -138,23 +109,39 @@ func Network() *schema.Resource {
...
+								"Pass `shared` to use this object across all the Groups.",
 						},
-						"vlan_id": {
-							Type:        schema.TypeString,
-							Optional:    true,
-							Description: "VLAN IDs eg. `0,3-5`. Use this field for VLAN based segments.",
+					},
+				},
+			},
+			"network_proxy": {
```
Comment (manjunath-batakurki) [inline internal/resources/resource_networks.go:133]: let proxy_id be as it was earlier

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+		},
+		ConflictsWith: []string{
+			"dhcp_network",
+		},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"code": {
+					Type:        schema.TypeString,
+					Computed:    true,
+					Description: "Network Type code",
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:29]: why code, ids are here ? are they specific to static network ?

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 14]
```diff
@@ -0,0 +1,149 @@
...
+					MaxItems:    1,
+					Elem: &schema.Resource{
+						Schema: map[string]*schema.Schema{
+							"connected_gateway": {
+								Type:     schema.TypeString,
+								Optional: true,
+								Description: "Connected Gateway. Pass Provider ID of the Tier1 gateway. Use " + DSRouter +
+									".provider_id  here.",
+							},
+							"vlan_ids": {
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:111]: why connect gateway and vlan_ids are here ? are they not common among both the networks ?

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+								Type:        schema.TypeString,
+								Optional:    true,
+								Description: "VLAN IDs eg. `0,3-5`. Use this field for VLAN based segments.",
+							},
+							"dhcp_type": {
+								Type:        schema.TypeString,
+								Required:    true,
+								Description: `DHCP Server type. Supported Values are "dhcpLocal", "dhcpRelay", "gatewayDhcp"`,
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:119]: Can users specify anything other than dhcpLocal ? please check with Jinu.

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 16]
```diff
@@ -0,0 +1,149 @@
...
+							},
+							"dhcp_server": {
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:121]: how can they get dhcp server id ?

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+								Description: "DHCP Server address and its CIDR. This address must not overlap the" +
+									"ip-ranges of the subnet, or the gateway address of the subnet," +
+									"or the DHCP static-binding addresses of this segment",
+							},
+							"dhcp_range": {
+								Type:        schema.TypeString,
+								Required:    true,
+								Description: "DHCP server IP Address range",
+							},
+							"dhcp_lease_time": {
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:138]: why lease time is string ?

File: internal/resources/resource_networks.go [hunk 18]
```diff
@@ -138,23 +109,39 @@ func Network() *schema.Resource {
 				Optional:    true,
 				Description: "If set to true, network will allow static override",
 			},
-			"config": {
+			"scope_id": {
```
Comment (manjunath-batakurki) [inline internal/resources/resource_networks.go:112]: why is pool_id outside ? it should be under static_network config right ?

File: examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf (hunk 2 above)
Comment (Pushpalatha555) [reply examples/resources/hpegl_vmaas_network/nsx_t_dhcp_network.tf]: updated

File: internal/cmp/resource_network.go (hunk 4 above)
Comment (Pushpalatha555) [reply internal/cmp/resource_network.go]: yes its working correctly

File: internal/cmp/resource_network.go (hunk 5 above)
Comment (Pushpalatha555) [reply internal/cmp/resource_network.go:144]: updated

File: internal/cmp/resource_network.go (hunk 6 above)
Comment (Pushpalatha555) [reply internal/cmp/resource_network.go:163]: updated

File: internal/resources/data_source_dhcp_server.go (hunk 7 above)
Comment (Pushpalatha555) [reply internal/resources/data_source_dhcp_server.go:34]: done

File: internal/resources/diffValidation/resource_nsxt_network.go (hunk 8 above)
Comment (Pushpalatha555) [reply internal/resources/diffValidation/resource_nsxt_network.go:56]: updated

File: internal/resources/resource_networks.go (hunk 9 above)
Comment (Pushpalatha555) [reply internal/resources/resource_networks.go]: updated

File: internal/resources/resource_networks.go (hunk 11 above)
Comment (Pushpalatha555) [reply internal/resources/resource_networks.go]: updated

File: internal/resources/resource_networks.go (hunk 12 above)
Comment (Pushpalatha555) [reply internal/resources/resource_networks.go]: updated

File: internal/resources/schemas/nsx_t_network_schemas.go (hunk 13 above)
Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go]: no updated

File: internal/resources/schemas/nsx_t_network_schemas.go (hunk 14 above)
Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go]: common for both updated

File: internal/resources/schemas/nsx_t_network_schemas.go (hunk 15 above)
Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go]: updated only dhcpLocal for now

File: internal/resources/schemas/nsx_t_network_schemas.go (hunk 16 above)
Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go]: updated the datasource to obtain it

File: internal/resources/schemas/nsx_t_network_schemas.go (hunk 17 above)
Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go]: its string only

File: internal/resources/resource_networks.go (hunk 18 above)
Comment (Pushpalatha555) [reply internal/resources/resource_networks.go:138]: yes updated

File: internal/resources/schemas/nsx_t_network_schemas.go [hunk 19]
```diff
@@ -0,0 +1,82 @@
...
+			"dhcp_network",
+		},
+		ConflictsWith: []string{
+			"dhcp_network",
+		},
+		Elem: &schema.Resource{
+			Schema: map[string]*schema.Schema{
+				"pool_id": {
+					Type:        schema.TypeInt,
+					Optional:    true,
```
Comment (manjunath-batakurki) [inline internal/resources/schemas/nsx_t_network_schemas.go:29]: when pool_id is optional, do you need to check if the data is empty inside this ? which you are doing in diff validate..

These changes may break the acceptance test cases i guess.

Comment (Pushpalatha555) [reply internal/resources/schemas/nsx_t_network_schemas.go:29]: updated

File: internal/resources/diffValidation/resource_nsxt_network.go [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+
+	return nil
+}
+
+func (l *Network) validateNetworkConfigs(networkTypes string) error {
+	value := l.diff.Get(networkTypes)
+	if len((value).([]interface{})) != 0 {
+		return fmt.Errorf("please provide " + staticNetwork + " for the Configuration")
```
[the remark this answers is not in the document]
Comment (Pushpalatha555) [reply internal/resources/diffValidation/resource_nsxt_network.go]: updated

**Review by manjunath-batakurki — APPROVED** [review]

**Review by ravikumar29524 — APPROVED** [review]
